### ISSUE TITLE
Custom scalars not coerced if used as mutation input

### ISSUE THREAD

**Issue by Fruko** [Netflix/dgs-framework#49]
Hey guys,

First of all, I really like your framework! One thing that bugs me though, it seems that custom scalars are not coerced in the mutation. As per documentation the `@InputArgument` can be used in mutations see [docs](https://netflix.github.io/dgs/mutations/)

Give the schema with mutation such as
```
type Mutation {
    scalarAsInput(scalar: [UUID]): [UUID]
}

scalar UUID
```

and registering the UUIDScalar using the `@DgsScalar(name="UUID")` and finally writing a DataFetcher 

```
@DgsComponent
public class ScalarDataFetcher {
    @DgsData(parentType = DgsConstants.MUTATION.TYPE_NAME, field = DgsConstants.MUTATION.ScalarAsInput)
    public List<UUID> customScalarMutation(@InputArgument("scalar") List<UUID> scalar) {
        return scalar;
    }
}
```

I am getting the following error
```
    "errors": [
        {
            "message": "Can't serialize value (/scalarAsInput[0]) : Not a valid UUID",
            "path": [
                "scalarAsInput",
                0
            ],
            "extensions": {
                "classification": "DataFetchingException"
            }
        }
    ],
```

When debugging I can see that the `DataFetchingEnvironment` has the argument properly coerced from `String` to `UUID`

![image](https://user-images.githubusercontent.com/10186522/107389085-bff5fb00-6af6-11eb-846f-9df7ef64b335.png)

I forked the examples repository to reproduce the issue [here](https://github.com/Fruko/dgs-examples-java)

**Comment by srinivasankavitha** [comment]
Thanks for the detailed description! We will look into fixing it.

**Comment by srinivasankavitha** [comment]
After looking into this some more, it appears that the handling of the custom scalar,i .e., just returning UUID actually works. However, List<UUID> does not because graphql-java unboxes the List type and treats the individual element as a String. In the implementation of the UUID scalar, the serialize method currently only checks for instanceof UUID (while it is a string in this case)
```
@Override
    public String serialize(Object dataFetcherResult) throws CoercingSerializeException {
        if (dataFetcherResult instanceof UUID) {
            return dataFetcherResult.toString();
        } else {
            throw new CoercingSerializeException("Not a valid UUID");
        }
    }
```
If we modify serialize() to also handle string, it works correctly:
```
@Override
    public String serialize(Object dataFetcherResult) throws CoercingSerializeException {
        if (dataFetcherResult instanceof UUID || dataFetcherResult instanceof String) {
            return dataFetcherResult.toString();
        } else {
            throw new CoercingSerializeException("Not a valid UUID");
        }
    }
```
This is a good use case that we haven't documented, so thanks for reporting.

**Comment by Fruko** [comment]
@srinivasankavitha thanks for quick reply and explanation! However, I applied the changes you suggested, it correctly changes the return type to UUID. However the type of `@InputArgument("scalar")`  remains String.

I have modified my fork to throw an exception to show the issue like this 
```
@DgsComponent
public class ScalarDataFetcher {
    @DgsData(parentType = DgsConstants.MUTATION.TYPE_NAME, field = DgsConstants.MUTATION.ScalarAsInput)
    public List<UUID> customScalarMutation(@InputArgument("scalar") List<UUID> scalar,
                                           DataFetchingEnvironment dataFetchingEnvironment) throws Exception {
        if(null != scalar.get(0) && scalar.get(0) instanceof UUID) {
            return scalar;
        } else {
            throw new Exception("Not UUID");
        }
    }
}
```
And when I send the mutation like this you can see the exception
![image](https://user-images.githubusercontent.com/10186522/107431203-686f8380-6b26-11eb-834c-051a957f7c8e.png)

So perhaps I am mistaken about how it should work, but the argument `@InputArgument("scalar") List<UUID> scalar` shouldn't be `List<String> scalar`

**Comment by srinivasankavitha** [comment]
Ok, you're right. I misunderstood because when I tried it as a simple UUID (not List<UUID>), the type was correctly identified as UUID. Using a List, I was getting issues in returning the response, so suggesting adding the string handling. I see the problem you are pointing out now.

**Comment by srinivasankavitha** [comment]
@Fruko - Merged a fix for the reported issue. Will be available in the next release.

**Comment by srinivasankavitha** [comment]
Fixed in v3.4.1.

**Comment by Fruko** [comment]
tested in my fork and it works as expected, thanks!
